tutorials/git-advanced.rst

```diff
@@ -0,0 +1,604 @@
+    $ git push --set-upstream origin update-buffer-size
+    Total 0 (delta 0), reused 0 (delta 0)
+    remote:
+    remote: Create a pull request for 'update-buffer-size' on GitHub by visiting:
+    remote:      https://github.com/GITHUB_USERNAME/uchicago-cs-git-tutorial/pull/new/update-buffer-size
+    remote:
+    To git@github.com:GITHUB_USERNAME/uchicago-cs-git-tutorial.git
+     * [new branch]      update-buffer-size -> update-buffer-size
+    Branch 'update-buffer-size' set up to track remote branch 'update-buffer-size' from 'origin'.
+
+
+Now, let's switch to the ``main`` branch::
+
+    git checkout main
+
+The buffer size in ``echo.c`` will still be 500. Let's make a change
+that will conflict with the change we made on a separate branch:
+change the buffer to ``250``, and commit that change::
+
+    $ git add echo.c
+    $ git commit -m"Updated buffer size to 250"
+    [main e5ec414] Updated buffer size to 250
+     1 file changed, 2 insertions(+), 2 deletions(-)
+
+So, we're now in a situation where the ``main`` branch and the ``update-buffer-size``
+branches have each made conflicting changes to the same lines. While we've artificially
+created this situation, take into account that this can easily happen in a project
+with multiple developers: one developer could decide to change the buffer size to
+1000, while another decides that 250 is enough.
+
+Let's see what happens if we try to merge ``update-buffer-size``:
+
+::
+
+    $ git merge update-buffer-size
+    Auto-merging echo.c
+    CONFLICT (content): Merge conflict in echo.c
+    Automatic merge failed; fix conflicts and then commit the result.
+
+Git has detected a merge conflict! A merge conflict is basically Git telling us
+“I see two conflicting changes on these two branches, and I don’t know
+how to resolve them automatically; you need to tell me how to do it”.
+
+Merge conflicts will also be shown on ``git status``:
+
+::
+
+    $ git status
+    On branch main
+    Your branch is ahead of 'origin/main' by 2 commits.
+      (use "git push" to publish your local commits)
+
+    You have unmerged paths.
+      (fix conflicts and run "git commit")
+      (use "git merge --abort" to abort the merge)
+
+    Unmerged paths:
+      (use "git add <file>..." to mark resolution)
+        both modified:   echo.c
+
+    no changes added to commit (use "git add" and/or "git commit -a")
+
+
+If you open ``echo.c``, you will see something like this:
+
+::
+
+    int main(){
+    <<<<<<< HEAD
+        char input[250];
+        while(fgets(input, 250, stdin)){  //read from STDIN (aka command-line)
+    =======
+        char input[1000];
+        while(fgets(input, 1000, stdin)){  //read from STDIN (aka command-line)
+    >>>>>>> update-buffer-size
+            printf("%s\n", input);  //print out what user typed in
+            memset(input, 0, strlen(input));  //reset string to all 0's
+        }
+        return 1;
+    }
+
+In general, any time you see something like this:
```

Comment:
It would be a good idea to include a mention of how modern IDEs give you the option to select which changes to keep with a button as opposed to manually having to update the file.

Reply:
Done.